components/browser/icons/src/main/java/mozilla/components/browser/icons/processor/ColorProcessor.kt

```diff
@@ -0,0 +1,33 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.icons.processor
+
+import android.content.Context
+import android.graphics.Color.rgb
+import androidx.palette.graphics.Palette
+import mozilla.components.browser.icons.Icon
+import mozilla.components.browser.icons.IconRequest
+
+/**
+ * [IconProcessor] implementation to extract the dominant color from the icon.
+ */
+class ColorProcessor : IconProcessor {
+
+    override fun process(context: Context, request: IconRequest, resource: IconRequest.Resource?, icon: Icon): Icon {
+        if (icon.bitmap == null || icon.color != null) return icon
+
+        val dominantColor = Palette.from(icon.bitmap).generate().getDominantColor(DEFAULT_COLOR)
+        return icon.copy(color = dominantColor and DOMINANT_COLOR_MASK)
+    }
+
+    companion object {
+        /**
+         * No color, here we use photon color
+         */
+        internal val DEFAULT_COLOR = rgb(177, 177, 179)
+
+        private const val DOMINANT_COLOR_MASK = 0x7FFFFFFF
```

Comment:
Here I also wonder if we should leave that up to the app since the app may be opinionated on how it wants to use the color.